Add storage capacity scoring KEP (#1)
topolvm/enhancements: keps/sig-storage/0000-storage-capacity-scoring-for-dynamic-provisioning/README.md

  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
- [ ] (R) Graduation criteria is in place
  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
- [ ] (R) Production readiness review completed
- [ ] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

> This KEP proposes adding dynamic provisioning scores using storage capacity to scoring with the VolumeBinding plugin.

Reviewer: ```suggestion
This KEP proposes adding scoring nodes for dynamic provisioning PVs using storage capacity in the VolumeBinding plugin.
```

> By considering the free space for dynamic provisioning, it is possible to schedule pods considering storage capacity.

Reviewer: ```suggestion
By considering the free spaces of nodes for dynamic provisioning, it is possible to schedule pods on the node that has the most or least free space.
```

<!--
This section is incredibly important for producing high-quality, user-focused
documentation such as release notes or a development roadmap. It should be
possible to collect this information before implementation begins, in order to
avoid requiring implementors to split their attention between writing release
notes and implementing the feature itself. KEP editors and SIG Docs
should help to ensure that the tone and content of the `Summary` section is
useful for a wide audience.

A good summary is probably at least a paragraph in length.

Both in this section and below, follow the guidelines of the [documentation
style guide]. In particular, wrap lines to a reasonable length, to make it
easier for reviewers to cite specific portions, and to minimize diff churn on
updates.

[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
-->

## Motivation

On the following cases, it is needed to schedule pods considering storage capacity.

> - We want to select a node with as much free space as possible so that we can resize after a node-local PV is scheduled, and use storage as evenly as possible among the nodes.

Reviewer: > use storage as evenly as possible among the nodes

このケースがなぜ必要かがわからないので消したほうが良いかも。

Reviewer: どちらかというと resize したいから、フリースペースがあるノードを選びたいので、
Because we want to resize ..., it should select a node ... とかのほうが良いかも

Reviewer 2: 77cd66b で修正済み

- Or, we want to select a node with less free node space to reduce the number of nodes as much as possible.

<!--
This section is for explicitly listing the motivation, goals, and non-goals of
this KEP.  Describe why the change is important and the benefits to users. The
motivation section can optionally provide links to [experience reports] to
demonstrate the interest in a KEP within the wider Kubernetes community.

[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
-->

### Goals

> By considering the free space of Dynamic Provisioning, it is possible to schedule according to the free space of the volume.

Reviewer: ゴールなので以下のような感じのほうがよさそうです。

```suggestion
Considering free spaces of nodes for dynamic provisioning.
```

<!--
List the specific goals of the KEP. What is it trying to achieve? How will we
know that this has succeeded?
-->

### Non-Goals

> - Change the way its score with static and dynamic

Reviewer: ```suggestion
- Change how to score nodes for static or dynamic provisioning.
```

<!--
What is out of scope for this KEP? Listing non-goals helps to focus discussion
and make progress.
-->

## Proposal

> - Dynamic Provisioning can be scored based on available space

Reviewer: dynamic positioning 自体はスコアを持っているわけではないので、以下のような感じでしょうか

Score of nodes based on available space can be taken into account when dynamic provisioning.

Reviewer 2: 77cd66b で修正済み

<!--
This is where we get down to the specifics of what the proposal actually is.
This should have enough detail that reviewers can understand exactly what
you're proposing, but should not include things like API designs or
implementation. What is the desired outcome and how do we measure success?.
The "Design Details" section below is for the real
nitty-gritty.
-->

### User Stories (Optional)

<!--
Detail the things that people will be able to do if this KEP is implemented.
Include as much detail as possible so that people can understand the "how" of
the system. The goal here is to make this feel real for users without getting
bogged down.
-->

#### Story 1

> If we want to leave room for volume expansion after node allocation, we can provision with free space in consideration during Dynamic Provisioning, so that nodes with the maximum amount of free space required can be prioritized.

Reviewer: > we can provision with free space in consideration during Dynamic Provisioning, so that nodes with the maximum amount of free space required can be prioritized.

この辺りの意味が読み取れませんでした。

Reviewer 2: d95b701 で修正済み

#### Story 2

> When using the cloud, we want to reduce the number of nodes as much as possible to reduce costs. Therefore, we would like to increase the priority of nodes with as little remaining capacity as possible when scheduling pods so as not to create useless nodes as much as possible. By making it possible to provision with free space in mind during Dynamic Provisioning, it is possible to preferentially allocate nodes with the minimum necessary free space.

Reviewer: ```suggestion
When using the cloud environment, we want to reduce the number of nodes as much as possible to reduce costs. Therefore, we would like to increase the priority of nodes with as little remaining capacity as possible when scheduling pods so as not to create useless nodes as much as possible. By making it possible to provision with free space in mind during Dynamic Provisioning, it is possible to preferentially allocate nodes with the minimum necessary free space.
```

so as not to create useless nodes as much as possible この辺の意味が良く読み取れませんでした。

> By making it possible to provision with free space in mind during Dynamic Provisioning, it is possible to preferentially allocate nodes with the minimum necessary free space.

カンマの前後で同じこと言っているだけに見えるので、文章構造を変えたほうがよさそうです。

Reviewer 2: d95b701 で修正済み

### Notes/Constraints/Caveats (Optional)

<!--
What are the caveats to the proposal?
What are some important details that didn't come across above?
Go in to as much detail as necessary here.
This might be a good place to talk about core concepts and how they relate.
-->

## Design Details

<!--
This section should contain enough information that the specifics of your
change are understandable. This may include API specs (though not always
required) or even code snippets. If there's any ambiguity about HOW your
proposal will be implemented, this is the place to discuss them.
-->

> We modify the existing VolumeBinding plugin to achieve Dynamic Provisioning scoring.

Reviewer: ```suggestion
We modify the existing VolumeBinding plugin to achieve scoring for Dynamic Provisioning.
```

### Modify stateData to be able to store StorageCapacity

> We modify the struct of `PodVolumes` contained in `stateData` for Dynamic Provisioning scoring.

Reviewer: ```suggestion
We modify the struct of `PodVolumes` contained in `stateData` to score nodes for Dynamic Provisioning.
```

- 		return true, nil
+ 		return true, nil, nil
	}

	// Look for a matching CSIStorageCapacity object(s).
	// TODO (for beta): benchmark this and potentially introduce some kind of lookup structure (https://github.com/kubernetes/enhancements/issues/1698#issuecomment-654356718).
	capacities, err := b.csiStorageCapacityLister.List(labels.Everything())
	if err != nil {
- 		return false, err
+ 		return false, nil, err
	}

  sizeInBytes := quantity.Value()
	for _, capacity := range capacities {
		if capacity.StorageClassName == storageClass.Name &&
			capacitySufficient(capacity, sizeInBytes) &&
			b.nodeHasAccess(node, capacity) {
			// Enough capacity found.
- 			return true, nil
+ 			return true, capacity, nil
		}
	}

	// TODO (?): this doesn't give any information about which pools where considered and why
	// they had to be rejected. Log that above? But that might be a lot of log output...
	klog.V(4).InfoS("Node has no accessible CSIStorageCapacity with enough capacity for PVC",
		"node", klog.KObj(node), "PVC", klog.KObj(claim), "size", sizeInBytes, "storageClass", klog.KObj(storageClass))
- 	return false, nil
+ 	return false, nil, nil
}
```

### Scoring for Dynamic Provisioning

> Add scoring with Dynamic Provisioning in the Score method of the VolumeBinding plugin. The scoring targets are each entry in `podVolumes.DynamicProvisions` where `Capacity` is not equal to `nil`.

Reviewer: ```suggestion
Add scoring with Dynamic Provisioning in the `Score` method of the VolumeBinding plugin. The scoring targets are each entry in `podVolumes.DynamicProvisions` where `Capacity` is not equal to `nil`.
```

Scoring is implemented using the existing StaticBindings mechanism. Set the followings to `classResources` passed to the `scorer` function:

- `Requested: provision.PVC.Spec.Resources.Requests[v1.ResourceName(v1.ResourceStorage)]`
- `Capacity: CSIStorageCapacity`

> By doing that, based on the `Shape` setting of `VolumeBindingArgs`, scoring that takes into account the free space of Dynamic Provisioning is realized.

Reviewer: ```suggestion
By doing that, based on the `Shape` setting of `VolumeBindingArgs`, scoring that takes into account the free space of Dynamic Provisioning is achived.
```

```diff
// Score invoked at the score extension point.
func (pl *VolumeBinding) Score(ctx context.Context, cs *framework.CycleState, pod *v1.Pod, nodeName string) (int64, *framework.Status) {
	if pl.scorer == nil {
		return 0, nil
	}
	state, err := getStateData(cs)
	if err != nil {
		return 0, framework.AsStatus(err)
	}
	podVolumes, ok := state.podVolumesByNode[nodeName]

    ...

+   // add dynamic binding volumes by storage class
+ 	for _, provision := range podVolumes.DynamicProvisions {
+ 		if provision.Capacity == nil {
+ 			continue
+ 		}
+ 		class := *provision.PVC.Spec.StorageClassName
+ 		if _, ok := classResources[class]; !ok {
+ 			classResources[class] = &StorageResource{
+ 				Requested: 0,
+ 				Capacity:  0,
+ 			}
+ 		}
+ 		requestedQty := provision.PVC.Spec.Resources.Requests[v1.ResourceName(v1.ResourceStorage)]
+ 		classResources[class].Requested += requestedQty.Value()
+ 		classResources[class].Capacity += provision.Capacity.Capacity.Value()
+ 	}

  return pl.scorer(classResources), nil
```

The PR below is a PoC implementation of the above design.  

> https://github.com/bells17/kubernetes/pull/1

Reviewer: KEPから外部リポジトリ参照するのは微妙なんで消したほうが良い気がします。

Reviewer 2: 77cd66b で修正済み

to implement this enhancement.

##### Prerequisite testing updates

<!--
Based on reviewers feedback describe what additional tests need to be added prior
implementing this enhancement to ensure the enhancements have also solid foundations.
-->

##### Unit tests

<!--
In principle every added code should have complete unit test coverage, so providing
the exact set of tests will not bring additional value.
However, if complete unit test coverage is not possible, explain the reason of it
together with explanation why this is acceptable.
-->

<!--
Additionally, for Alpha try to enumerate the core package you will be touching
to implement this enhancement and provide the current unit coverage for those
in the form of:
- <package>: <date> - <current test coverage>
The data can be easily read from:
https://testgrid.k8s.io/sig-testing-canaries#ci-kubernetes-coverage-unit

This can inform certain test coverage improvements that we want to do before
extending the production code to implement this enhancement.
-->

The following unit tests are planned.

- Is the scoring for Dynamic Provisioning appropriate to free space?

> - Are the free space score for Dynamic Provisioning and the Static Bindings score both functionally?

Reviewer: ```suggestion
- Are the free space score for Dynamic Provisioning and the Static Bindings score both functional?
```
